### PR TITLE
Introduce Apply plan node for subqueries

### PR DESCRIPTION
**Pull request by kokosing**
It is a logical representation of a subquery.
    Apply node contains input relation, expression relation and parameters
    (symbol list).
    
    Each row in the input relation is passed to the expression, then
    expression (subquery) is executed. Expression produces new columns which
    are joined with the input row. Parameters (correlation list) determines
    what columns need to be passed from the input to the expression to
    calculate the result rows. If correlation list is empty then subquery
    is not correlated, in this case values returned from expression are
    always the same, not dependent on input relation.
    
    See
    https://docs.google.com/document/d/18HN7peS2eR8lZsErqcmnoWyMEPb6p4OQeidH1JP_EkA/

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 1]
```diff
@@ -634,41 +633,34 @@ private PlanBuilder appendProjections(PlanBuilder subPlan, Iterable<Expression>
...
     {
         for (InPredicate inPredicate : inPredicates) {
-            subPlan = appendSemiJoin(subPlan, inPredicate);
+            subPlan = appendInPredicateApply(subPlan, inPredicate);
         }
         return subPlan;
     }
 
-    private PlanBuilder appendSemiJoin(PlanBuilder subPlan, InPredicate inPredicate)
+    private PlanBuilder appendInPredicateApply(PlanBuilder subPlan, InPredicate inPredicate)
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: this code is identical with the one in QueryPlanner? Should I refactor it?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java [hunk 2]
```diff
@@ -266,8 +265,8 @@ else if (firstDependencies.stream().allMatch(right.canResolvePredicate()) && sec
 
             // Add semi joins if necessary
             if (joinInPredicates != null) {
```
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/RelationPlanner.java]: What are the joinInPredicates?

**Comment by cberner** [comment]
Looks like the tests are failing. Can you take a look at Travis?

**Comment by pnowojski** [comment]
@cberner @kokosing: travis fixed, top commit should to be squashed before merging with master.

**Comment by kokosing** [comment]
@cberner @martint ping

**Comment by cberner** [comment]
I'm working with @sopel39 on `EXPLAIN ANALYZE` at the moment. Can you remind me once that is done?

**Comment by kokosing** [comment]
@cberner @martint ping

**Comment by kokosing** [comment]
@martint Can you please review? This is base refactor needed to start the work on correlated subqueries.

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 3]
```diff
@@ -334,6 +336,9 @@ protected void setup(Binder binder)
         // optimizers
         binder.bind(new TypeLiteral<List<PlanOptimizer>>() {}).toProvider(PlanOptimizersFactory.class).in(Scopes.SINGLETON);
 
+        // plan sanity checkers
+        binder.bind(new TypeLiteral<List<PlanSanityChecker>>() {}).toProvider(PlanSanityCheckersFactory.class).in(Scopes.SINGLETON);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: I don't think this abstraction is necessary, especially since they don't have dependencies on other classes (like metadata, etc). I would just handle all this within PlanSanityChecker (which could hold a list of checkers internally and iterate over them in the `validate()` call`

**Comment by kokosing** [comment]
@martint comment applied and rebased

**Comment by martint** [comment]
I merged the first 5 commits.

**Comment by kokosing** [comment]
Thanks. I rebased.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoSubqueryExpressionLeftSanityChecker.java [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+    @Override
+    public void validate(PlanNode plan)
+    {
+        for (Expression expression : ExpressionExtractor.extractExpressions(plan)) {
+            new DefaultTraversalVisitor<Void, Void>()
+            {
+                @Override
+                protected Void visitSubqueryExpression(SubqueryExpression node, Void context)
+                {
+                    throw new SemanticException(SemanticErrorCode.NOT_SUPPORTED, node, "Given subquery is not yet supported");
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/NoSubqueryExpressionLeftSanityChecker.java]: Plan sanity checkers are about enforcing invariants. I.e., make sure optimizers are not breaking assumptions and producing incorrect plans. So, they should throw IllegalArgumentExceptions instead of SemanticExceptions (the analyzer should be throwing these)

File: presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java [hunk 5]
```diff
@@ -156,10 +156,10 @@ protected void setup(Binder binder)
 
         if (serverConfig.isCoordinator()) {
             discoveryBinder(binder).bindHttpAnnouncement("presto-coordinator");
-            binder.bind(new TypeLiteral<Optional<QueryPerformanceFetcher>>(){}).toProvider(QueryPerformanceFetcherProvider.class).in(Scopes.SINGLETON);
+            binder.bind(new TypeLiteral<Optional<QueryPerformanceFetcher>>() {}).toProvider(QueryPerformanceFetcherProvider.class).in(Scopes.SINGLETON);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/server/ServerMainModule.java]: Unrelated changes

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/EnsureSymbolDependenciesPlanSanityChecker.java [hunk 6]
```diff
@@ -63,11 +65,11 @@
 /**
  * Ensures that all dependencies (i.e., symbols in expressions) for a plan node are provided by its source nodes
  */
-public final class PlanSanityChecker
+public final class EnsureSymbolDependenciesPlanSanityChecker
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/EnsureSymbolDependenciesPlanSanityChecker.java]: Call this "ValidateDependencies"

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java [hunk 7]
```diff
@@ -0,0 +1,98 @@
...
+        extends PlanNode
+{
+    private final PlanNode input;
+    private final PlanNode subquery;
+
+    // Mapping between input and subquery, key represents symbols in input, values in subquery
+    private final Map<Symbol, Symbol> correlation;
+
+    @JsonCreator
+    public ApplyNode(@JsonProperty("id") PlanNodeId id,
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: Move argument to the next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java [hunk 8]
```diff
@@ -0,0 +1,98 @@
...
+
+@Immutable
+public class ApplyNode
+        extends PlanNode
+{
+    private final PlanNode input;
+    private final PlanNode subquery;
+
+    // Mapping between input and subquery, key represents symbols in input, values in subquery
+    private final Map<Symbol, Symbol> correlation;
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: I'm not sure I understand how this is supposed to work. The thing that confuses me is that it's linking the fields from the outer subplan to the outputs of the inner subplan, but in my mental model, the fields in the outer are supposed to be *inputs* to the inner that get "bound" on every iteration of the inner subquery (i.e., like arguments to a function call).

File: presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/PlanSanityChecker.java [hunk 9]
```diff
@@ -0,0 +1,38 @@
...
+import com.google.common.collect.ImmutableList;
+
+import java.util.List;
+
+/**
+ * It is going to be executed at the end of logical planner, to verify its correctness
+ */
+public interface PlanSanityChecker
+{
+    List<PlanSanityChecker> PLAN_SANITY_CHECKERS = ImmutableList.of(
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/sanity/PlanSanityChecker.java]: I think it's confusing to have this on the interface that the checkers implement. I'd have a PlanSanityChecker class with a single static method (and a static list of checkers), and a separate interface checkers implement. E.g.:

```java
class PlanSanityChecker
{
    public static void validate(PlanNode plan) { ... }

    public interface Checker 
    {
        void validate(PlanNode plan);
    }
}

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java [hunk 10]
```diff
@@ -0,0 +1,70 @@
...
+import com.facebook.presto.sql.planner.plan.PlanNode;
+import com.facebook.presto.sql.planner.plan.SimplePlanRewriter;
+import com.google.common.collect.ImmutableList;
+
+import java.util.Map;
+import java.util.Optional;
+
+import static java.util.Objects.requireNonNull;
+
+public class UncorrelatedScalarApplyRemover
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java]: I'd might rename this to TransformUncorrelatedScalarToJoin

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java [hunk 11]
```diff
@@ -0,0 +1,70 @@
...
+
+        public Rewriter(PlanNodeIdAllocator idAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitApply(ApplyNode node, RewriteContext<PlanNode> context)
+        {
+            node = (ApplyNode) super.visitApply(node, context);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java]: Call `context.defaultRewrite(...)` instead.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java [hunk 12]
```diff
@@ -0,0 +1,70 @@
...
+
+            if (node.getCorrelation().isEmpty() && node.getSubquery() instanceof EnforceSingleRowNode) {
+                // only scalar subquery wraps expression in EnforceSingleRowNode
+                return new JoinNode(idAllocator.getNextId(),
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedScalarApplyRemover.java]: Move argument to next line

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+import com.facebook.presto.sql.planner.PlanNodeIdAllocator;
+import com.facebook.presto.sql.planner.SimplePlanVisitor;
+import com.facebook.presto.sql.planner.Symbol;
+import com.facebook.presto.sql.planner.SymbolAllocator;
+import com.facebook.presto.sql.planner.plan.ApplyNode;
+import com.facebook.presto.sql.planner.plan.PlanNode;
+
+import java.util.Map;
+
+public class VerifyNoApplyNodeLeftPlanOptimizer
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java]: I'm not sure I like the idea of having this as a plan optimizer. Can we make it a sanity checker, instead?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java [hunk 14]
```diff
@@ -0,0 +1,281 @@
...
+        public InPredicateRewriter(PlanNodeIdAllocator idAllocator, SymbolAllocator symbolAllocator)
+        {
+            this.idAllocator = requireNonNull(idAllocator, "idAllocator is null");
+            this.symbolAllocator = requireNonNull(symbolAllocator, "symbolAllocator is null");
+        }
+
+        @Override
+        public PlanNode visitFilter(FilterNode node, RewriteContext<Void> context)
+        {
+            node = (FilterNode) super.visitFilter(node, context);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java]: Call `context.defaultRewrite(...)` instead

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java [hunk 15]
```diff
@@ -0,0 +1,281 @@
...
+            ImmutableMap.Builder<InPredicate, Expression> inPredicateMapping = ImmutableMap.builder();
+            node = rewriteInPredicates(node, inPredicates, inPredicateMapping);
+            Expression expression = replaceExpression(node.getPredicate(), inPredicateMapping.build());
+            return new FilterNode(node.getId(), getOnlyElement(node.getSources()), expression);
+        }
+
+        @Override
+        public PlanNode visitProject(ProjectNode node, RewriteContext<Void> context)
+        {
+            node = (ProjectNode) super.visitProject(node, context);
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java]: Same here

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java [hunk 16]
```diff
@@ -0,0 +1,281 @@
...
+        {
+            ImmutableMap.Builder<Symbol, Expression> assignmentsBuilder = ImmutableMap.builder();
+            Map<Symbol, Expression> assignments = node.getAssignments();
+            for (Symbol symbol : assignments.keySet()) {
+                assignmentsBuilder.put(symbol, replaceExpression(assignments.get(symbol), inPredicateMapping));
+            }
+            return assignmentsBuilder.build();
+        }
+
+        private List<InPredicate> collectInPredicates(Collection<Expression> expressions)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java]: I'd call this method `extractInPredicates`

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java [hunk 17]
```diff
@@ -0,0 +1,281 @@
...
+        {
+            ImmutableList.Builder<InPredicate> listBuilder = ImmutableList.builder();
+            for (Expression expression : expressions) {
+                new DefaultTraversalVisitor<Void, Void>()
+                {
+                    @Override
+                    protected Void visitInPredicate(InPredicate node, Void context)
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java]: This will also include any predicates of the form `xxx in (a, b, c, ...)`. Is that intentional? Isn't this supposed to do it only for cases where the "value expression" of the IN predicate is a subquery?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java [hunk 18]
```diff
@@ -0,0 +1,281 @@
...
+import java.util.Optional;
+
+import static com.facebook.presto.spi.type.BooleanType.BOOLEAN;
+import static com.facebook.presto.sql.planner.ExpressionNodeInliner.replaceExpression;
+import static com.facebook.presto.sql.planner.plan.SimplePlanRewriter.rewriteWith;
+import static com.google.common.base.Preconditions.checkState;
+import static com.google.common.collect.Iterables.getOnlyElement;
+import static java.util.Objects.requireNonNull;
+
+public class UncorrelatedInPredicateApplyRemover
```
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java]: Can you add an example to javadoc comments of whats kind of plan pattern it applies to and what the result looks like?

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java (hunk 13 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java]: It used to be written as sanity checker. 

Plan sanity checkers are run before optimizers where apply nodes still exist. I could change sanity checkers to accept parameter flag which would tell if plan is optimized or not, but to make this work all other optimizers would need to support Apply node. Otherwise, instead of getting an error from sanity checkers we could get some mysterious error from other optimizer rule.

I don't like it either but I don't have a better idea for this.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 8 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: I think your mental model is correct at least it is in line with my mental model. Maybe I didn't cast that into java code clearly enough. I think that comment above correlation map could be misleading, I will fix it.

The direction of mapping in correlation map to me does not have much importance. Would it be cleaner to you I would reverse it? In that case, key in map would represent some symbol in a subquery plan (inner plan) - an input to subquery call, and value in map would represent output symbol from input node (outer plan).

Take a look at the example of a real query:
`SELECT n.nationkey FROM nation n WHERE n.nationkey + 100 < (SELECT n.nationkey*2);`

This query could have below plan:
```
Project(nationkey) -> (nationkey)
  Filter(nationkey + 100 < expr) -> (nationkey, expr)
    Apply -> (nationkey, expr)
      (correlation) {nationkey: nationkey_1}
      (input) TableScan(nation) -> (nationkey)
      (subquery) EnforceSingleRow -> (expr)
        Project(expr -> nationkey_1 * 2) -> (expr)
```
At the above example, `input` of apply node returns `nationkey` symbol which is then used in `subquery` plan in place of `nationkey_1` symbol.

However, I have an idea of using plain list of correlation symbols instead of map. Then plan would like below:
```
Project(nationkey) -> (nationkey)
  Filter(nationkey + 100 < expr) -> (nationkey, expr)
    Apply -> (nationkey, expr)
      (correlation) [nationkey]
      (input) TableScan(nation) -> (nationkey)
      (subquery) EnforceSingleRow -> (expr)
        Project(expr -> nationkey * 2) -> (expr)
```

See that input and subquery plans are both using the same symbol `nationkey` (no need for `nationkey_1`). So Apply need to only know which symbols are correlated. I think it is easier to understand such plan. I will change that.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java (hunk 17 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/UncorrelatedInPredicateApplyRemover.java]: For cases of regular IN predicate - `xxx in (a, b, c, ...)` optimizer won't find any matching `Apply` node, so it does not influence the correctness (only performance suffers a bit).

At this point subquery expression is rewritten to apply node `xxx in (select...)` is in form of `xxx in expr`. Where `expr` is a symbol referring to apply node output. Therefore it could be difficult to properly distinguish IN predicate with subquery to IN predicate with other symbol `xxx in (yyy)`, where `yyy` could be a reference to a column for example.

However, I could optimize this to select IN predicates which have only one symbol in valueList.

**Comment by kokosing** [comment]
@martint I addressed all the comments and answered all the questions. Comments are addressed in separate commits, I will merge them just before the merge. I didn't rebase on origin/master because I wanted to preserve the discussion comments for a while.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java (hunk 13 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java]: We can just remove the call to validate the plan before the optimizers run.

File: presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java (hunk 8 above)
Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: Ok, that makes sense. But, then, isn't the following check in the `ApplyNode` constructor incorrect?

```java
checkArgument(subquery.getOutputSymbols().containsAll(correlation.values()), "Subquery does not contain symbols from correlation");
```

For your first example above, this would be checking that `expr` (output of the inner subquery) is in the values of the correlation map, which contains `nationkey`->`nationkey_1`

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/plan/ApplyNode.java]: Yes, it is incorrect. I noticed that and I removed it in subsequent commit. Please see: https://github.com/prestodb/presto/pull/4403/commits/d15cf99ca337665758cae48f50c952a85cf198f2#diff-727404f7967058fb13c2898cad1c3784R52

File: presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java (hunk 13 above)
Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java]: Well... I think that having sanity checks before optimizers run has its own value. Thanks to that you know if plan was broken before or after optimizers run.

I would prefer to leave it as it is, unless you have a strong opinion about it. Then I will change it according to your comment.

Comment (martint) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java]: Generally, the way this ends up being used is:

* Get a "broken plan error"
* Distill the query to a small reproducible case
* Fire up Presto in a debugger & trace. This commonly involves adding a call to validate() before/after each optimizer.

So, you almost always have to tweak the code to use track down issues, so I think it's acceptable to just validate at the end and leave it up to the developer to insert calls and enable/disable checkers as necessary while debugging.

Comment (kokosing) [inline presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/VerifyNoApplyNodeLeftPlanOptimizer.java]: Sounds ok. Thanks for explanation.

śr., 4.05.2016, 21:46 użytkownik Martin Traverso <notifications@github.com>
napisał:

**Comment by kokosing** [comment]
Superseded by #5202
